File: September/2026-09-05/16/16-3sum-closest.py
# 16-3sum-closest.cpp;array-2p;self(help for input part);leetcode

from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()

        closest_sum = float('inf')

        for i in range(len(nums) - 2):
            left = i + 1
            right = len(nums) - 1

            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]

                if abs(current_sum - target) < abs(closest_sum - target):
                    closest_sum = current_sum

                if current_sum < target:
                    left += 1
                elif current_sum > target:
                    right -= 1
                else:
                    return current_sum

        return closest_sum


nums = list(map(int, input("Enter the elements of the array: ").split()))
target = int(input("Enter the target: "))

solution = Solution()
result = solution.threeSumClosest(nums, target)

print("Closest Sum:", result)

File: September/2026-09-05/16/16-3sum-closest.cpp
// 16-3sum-closest.cpp;array-2p;self(help for lib);leetcode

#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
#include <cstdlib>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        int closest_sum = INT_MAX / 2;

        for (int i = 0; i < nums.size() - 2; ++i) {
            int left = i + 1;
            int right = nums.size() - 1;

            while (left < right) {
                int current_sum = nums[i] + nums[left] + nums[right];

                if (abs(current_sum - target) < abs(closest_sum - target)) {
                    closest_sum = current_sum;
                }

                if (current_sum < target) {
                    ++left;
                }
                else if (current_sum > target) {
                    --right;
                }
                else {
                    return current_sum;
                }
            }
        }

        return closest_sum;
    }
};

int main() {
    Solution solution;

    vector<int> nums = {-1, 2, 1, -4};
    int target = 1;

    int result = solution.threeSumClosest(nums, target);

    cout << "Closest Sum: " << result << endl;

    return 0;
}